docs/path-resolution.rst

```diff
@@ -0,0 +1,454 @@
+point to the new version.
+
+Here are the detailed rules governing the behaviour of remappings:
+
+#. **Remappings only affect the translation between import paths and source unit names.**
+
+   Source unit names added via other means cannot be remapped.
+   For example the paths you specify on the command-line and the ones in ``sources.urls`` in
+   Standard JSON are not affected.
+
+    .. code-block:: bash
+
+        solc /project=/contracts /project/contract.sol # source unit name: /project/contract.sol
+
+#. **Context and prefix must match source unit names, not import paths.**
+
+   - This means that you cannot remap ``./`` or ``../`` directly since they are replaced during
+     the translation to source unit name but you can remap the part of the name they are replaced
+     with:
+
+     .. code-block:: bash
+
+         solc ./=a /project=b /project/contract.sol
+
+     .. code-block:: solidity
+         :caption: /project/contract.sol
+
+         import "./util.sol" as util; // source unit name: b/util.sol
+
+   - You cannot remap base path or any other part of the path that is only added internally by an
+     import callback:
+
+     .. code-block:: bash
+
+         solc /project=/contracts /project/contract.sol --base-path /project
+
+     .. code-block:: solidity
+         :caption: /project/contract.sol
+
+         import "util.sol" as util; // source unit name: util.sol
+
+#. **Target is inserted directly into the source unit name and does not necessarily have to be a valid path.**
+
+   - It can be anything as long as the import callback can handle it.
+     In case of the Host Filesystem Loader this includes also relative paths.
+     When using the JavaScript interface you can even use URLs and abstract identifiers if
+     your callback can handle them.
+
+   - Remapping happens after relative imports have already been resolved into source unit names.
+     This means that targets starting with ``./`` and ``../`` have no special meaning and are
+     relative to the base path rather than to the location of the source file.
+
+   - Remapping targets are not normalized so ``@root=./a/b//`` will remap ``@root/contract.sol``
+     to ``./a/b//contract.sol`` and not ``a/b/contract.sol``.
+
+   - If the target does not end with a slash, the compiler will not add one automatically:
+
+     .. code-block:: bash
+
+         solc /project/=/contracts /project/contract.sol
+
+     .. code-block:: solidity
+         :caption: /project/contract.sol
+
+         import "/project/util.sol" as util; // source unit name: /contractsutil.sol
+
+#. **Context and prefix are patterns and matches must be exact.**
+
+   - ``a//b=c`` will not match ``a/b``.
+   - source unit names are not normalized so ``a/b=c`` will not match ``a//b`` either.
+   - Parts of file and directory names can match as well.
+     ``/newProject/con:/new=old`` will match ``/newProject/contract.sol`` and remap it to
```

Comment:
Oha! That looks really crazy! I really wonder whether this mechanism is really used.

Reply:
I doubt it is. But people using it should be aware of it and know to always include the final `/`. Otherwise the results can be pretty confusing :)